subsys/modbus/mb_rtu_client.c

```diff
@@ -0,0 +1,615 @@
+ *                         The Embedded Modbus Stack
+ *
+ *      Copyright 2003-2020 Silicon Laboratories Inc. www.silabs.com
+ *
+ *                   SPDX-License-Identifier: APACHE-2.0
+ *
+ * This software is subject to an open source license and is distributed by
+ *  Silicon Laboratories Inc. pursuant to the terms of the Apache License,
+ *      Version 2.0 available at www.apache.org/licenses/LICENSE-2.0.
+ */
+
+#include <string.h>
+#include <sys/byteorder.h>
+#include <mb_rtu_internal.h>
+
+#include <logging/log.h>
+LOG_MODULE_REGISTER(mb_rtu_c, CONFIG_MODBUS_RTU_LOG_LEVEL);
+
+static int mbm_validate_response_fc(struct mb_rtu_context *pch,
+				    const uint8_t node_addr,
+				    uint8_t fc)
+{
+	uint8_t resp_fc = pch->rx_frame.fc;
+	uint8_t excep_code = pch->rx_frame.data[0];
+	const uint8_t excep_bit = BIT(7);
+	const uint8_t excep_mask = BIT_MASK(7);
+
+	if (node_addr != pch->rx_frame.addr) {
+		return -EIO;
+	}
+
+	if (fc != (resp_fc & excep_mask)) {
+		return -EIO;
+	}
+
+	if (resp_fc & excep_bit) {
+		if (excep_code > MODBUS_EXC_NONE) {
+			return excep_code;
+		}
+
+		return -EIO;
+	}
+
+	return 0;
+}
+
+#ifdef CONFIG_MODBUS_RTU_FP_EXTENSIONS
+static int mbm_validate_fc03fp_response(struct mb_rtu_context *pch, float *ptbl)
+{
+	uint8_t resp_byte_cnt;
+	uint8_t req_byte_cnt;
+	uint16_t req_qty;
+	uint8_t *resp_data;
+
+	resp_byte_cnt = pch->rx_frame.data[0];
+	resp_data = &pch->rx_frame.data[1];
+	req_qty = sys_get_be16(&pch->tx_frame.data[2]);
+	req_byte_cnt = req_qty * sizeof(float);
+
+	if (req_byte_cnt != resp_byte_cnt) {
+		LOG_ERR("Mismatch in the number of registers");
+		return -EINVAL;
+	}
+
+	for (uint8_t i = 0; i < req_qty; i++) {
+		uint32_t reg_val = sys_get_be32(resp_data);
+
+		memcpy(&ptbl[i], &reg_val, sizeof(float));
+		resp_data += sizeof(uint32_t);
+	}
+
+	return 0;
+}
+#endif
+
+static int mbm_validate_rd_response(struct mb_rtu_context *pch,
+				    const uint8_t node_addr,
+				    uint8_t fc,
+				    uint8_t *data)
+{
+	int err;
+	uint8_t resp_byte_cnt;
+	uint8_t req_byte_cnt;
+	uint16_t req_qty;
+	uint16_t req_addr;
+	uint8_t *resp_data;
+	uint16_t *data_p16 = (uint16_t *)data;
+
+	if (data == NULL) {
+		return -EINVAL;
+	}
+
+	resp_byte_cnt = pch->rx_frame.data[0];
+	resp_data = &pch->rx_frame.data[1];
+	req_qty = sys_get_be16(&pch->tx_frame.data[2]);
+	req_addr = sys_get_be16(&pch->tx_frame.data[0]);
+
+	switch (fc) {
+	case MODBUS_FC01_COIL_RD:
+	case MODBUS_FC02_DI_RD:
+		req_byte_cnt = (uint8_t)((req_qty - 1) / 8) + 1;
+		if (req_byte_cnt != resp_byte_cnt) {
+			LOG_ERR("Mismatch in the number of coils or inputs");
+			err = -EINVAL;
+		} else {
+			for (uint8_t i = 0; i < resp_byte_cnt; i++) {
+				data[i] = resp_data[i];
+			}
+
+			err = 0;
+		}
+		break;
+
+	case MODBUS_FC03_HOLDING_REG_RD:
+		if (IS_ENABLED(CONFIG_MODBUS_RTU_FP_EXTENSIONS) &&
+		    (req_addr >= MODBUS_RTU_FP_ADDR)) {
+			err = mbm_validate_fc03fp_response(pch, (float *)data);
+			break;
+		}
```

Comment:
```suggestion
#ifdef CONFIG_MODBUS_RTU_FP_EXTENSIONS
		if (req_addr >= MODBUS_RTU_FP_ADDR) {
			err = mbm_validate_fc03fp_response(pch, (float *)data);
			break;
		}
#endif
```
If CONFIG_MODBUS_RTU_FP_EXTENSIONS is not set mbm_validate_fc03fp_response is not compiled and therefore the compiler throws an implicit declaration warning.

Reply:
Any opinion on this?

Reply:
I will tackle it later this week, but I want to avoid the use of _ifdef_.

Reply:
I guessed that too and I think this is not a runtime issue, but having implicit defined function is nothing I want to have in production code :-D.

Reply:
Fixed by removing ifdef/endif arround mbm_validate_fc03fp_response().